azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -934,7 +934,9 @@ public void cancelExecutingJobs(ExecutableFlow exFlow, String userId,
   @Override
   public String submitExecutableFlow(ExecutableFlow exflow, String userId)
     throws ExecutorManagerException {
-    synchronized (exflow) {
+
+    String exFlowKey = exflow.getProjectName() + "." + exflow.getId() + ".submitFlow";
+    synchronized (exFlowKey.intern()) {
```

Comment:
@suvodeep-pyne can you elaborate your thoughts? Why do you think it's hacky and less obvious? An alternate solution I can think of is to have a hashmap whose key is combination of project name and flow name, value is its associated lock. But the implementation is more complicated.